code/modules/client/preference_setup/controls/01_keybindings.dm

```diff
@@ -0,0 +1,254 @@
+		window.location=url;
+		deedDone = true;
+	}
+	document.getElementById('focus').focus();
+	</script>
+	"}
+	winshow(user, "capturekeypress", TRUE)
+	var/datum/browser/popup = new(user, "capturekeypress", "<div align='center'>Keybindings</div>", 350, 300)
+	popup.set_content(HTML)
+	popup.open(FALSE)
+
+/datum/category_item/player_setup_item/controls/keybindings/OnTopic(href, list/href_list, mob/user)
+	switch(href_list["preference"])
+		if("keybindings_capture")
+			var/datum/keybinding/kb = global.keybindings_by_name[href_list["keybinding"]]
+			var/old_key = href_list["old_key"]
+			capture_keybinding(user, kb, old_key)
+			return TOPIC_REFRESH
+
+		if("keybindings_set")
+			var/kb_name = href_list["keybinding"]
+			if(!kb_name)
+				show_browser(user, null, "window=capturekeypress")
+				return TOPIC_REFRESH
+
+			var/clear_key = text2num(href_list["clear_key"])
+			var/old_key = href_list["old_key"]
+			if(clear_key)
+				if(pref.key_bindings[old_key])
+					pref.key_bindings[old_key] -= kb_name
+					if(!(kb_name in pref.key_bindings["None"]))
+						LAZYADD(pref.key_bindings["None"], kb_name)
+					if(!length(pref.key_bindings[old_key]))
+						pref.key_bindings -= old_key
+				show_browser(user, null, "window=capturekeypress")
+				user.client.set_macros()
+				SScharacter_setup.queue_preferences_save(pref)
+				return TOPIC_REFRESH
+
+			var/new_key = uppertext(href_list["key"])
+			var/AltMod = text2num(href_list["alt"]) ? "Alt" : ""
+			var/CtrlMod = text2num(href_list["ctrl"]) ? "Ctrl" : ""
+			var/ShiftMod = text2num(href_list["shift"]) ? "Shift" : ""
+			var/numpad = text2num(href_list["numpad"]) ? "Numpad" : ""
+
+			if(!new_key) // Just in case (; - not work although keyCode 186 and nothing should break)
+				show_browser(user, null, "window=capturekeypress")
+				return TOPIC_REFRESH
+
+			if(global._kbMap[new_key])
+				new_key = global._kbMap[new_key]
+
+			var/full_key
+			switch(new_key)
+				if("Alt")
+					full_key = "[new_key][CtrlMod][ShiftMod]"
+				if("Ctrl")
+					full_key = "[AltMod][new_key][ShiftMod]"
+				if("Shift")
+					full_key = "[AltMod][CtrlMod][new_key]"
+				else
+					full_key = "[AltMod][CtrlMod][ShiftMod][numpad][new_key]"
+			if(kb_name in pref.key_bindings[full_key]) //We pressed the same key combination that was already bound here, so let's remove to re-add and re-sort.
+				pref.key_bindings[full_key] -= kb_name
+			if(pref.key_bindings[old_key])
+				pref.key_bindings[old_key] -= kb_name
+				if(!length(pref.key_bindings[old_key]))
+					pref.key_bindings -= old_key
+			pref.key_bindings[full_key] += list(kb_name)
+			pref.key_bindings[full_key] = sortTim(pref.key_bindings[full_key], /proc/cmp_text_asc)
+
+			show_browser(user, null, "window=capturekeypress")
+			user.client.set_macros()
+			SScharacter_setup.queue_preferences_save(pref)
```

Comment:
Again, not good to save without the user actually hitting save. Not going to comment in other places.